### PR TITLE
Auto insert on reload

### PR DESCRIPTION
**Pull request by oliver-sanders**
**Not for merge**

Addresses #774 

This is a messy, part-baked implementation of auto-insert for tasks added through a suite reload:
- It determines whether a task can be inserted by forming a graph leading back to tasks currently in the task pool and assessing their status.
- The cycle point to insert the new task is that of the task pool task(s) that it leads to / hangs off adjusted for inter-cycle dependence extracted from config.edges (i.e. `foo[-PT6H] => bar`)
- If the new task is a graph isolate (its sub graph does not connect to anything from pre-reload) then the cycle point is taken to be the earliest point on the tasks sequence at or after the oldest active cycle point.
- If a task downstream of a new task has completed then the new task **will not** be inserted _(at this cycle point)_
- If a task upstream of a new task has succeeded then the new task **will** be inserted
- If a task upstream of a new task has failed and the dependency relationship is for a failed state (`foo:fail => bar`) then task **will** be inserted, else it **will not**.

I've been testing with the following suite:

``` ini
[scheduling]
    initial cycle point = 2000-01-01T00
    [[dependencies]]
        [[[T12]]]
            graph = d[-P1D] => a => b => c:fail => d
[runtime]
    [[root]]
        script = sleep 1
    [[a, b, d]]
    [[c]]
        script = fail
```

Reloading with:

``` ini
[scheduling]
    initial cycle point = 2000-01-01T00
    [[dependencies]]
        [[[T12]]]
            graph = """
                d[-P1D] => a => b => c:failed => d
                new_1 => a => new_2
                new_3 => new_4 & new_5
                new_4 & new_5 => b
                b => new_6 & new_7
                new_6 & new_7 => new_8
                new_9 => c
                c:fail => new_10
                new_11[-PT6H] => d
            """
        [[[T06]]]
            graph = new_11
        [[[T18]]]
            graph = d[-PT6H] => new_12
        [[[T00]]]
            graph = new_13 => new_14
[runtime]
    [[root]]
        script = sleep 1
    [[a, b, d]]
    [[c]]
        script = fail
```

@hjoliver @arjclark:
- Is this behaviour what you would expect?
- Should we auto insert by default or should it be a command line option?

### CODE REVIEW

**Comment by hjoliver** [comment]
@oliver-sanders - 

> Is this behaviour what you would expect?

This is quite complicated and I probably won't have time to consider it in detail this week (HPC benchmark prep...), but at first glance it seems to do the right thing.  To make my life easier, can you explain in what sense exactly this is "part-baked": is there some circumstance in which you think it doesn't do the right thing, or is it just that you want to tidy up the implementation?

> Should we auto insert by default or should it be a command line option?

I think this depends on how often (if any) "unexpected" automatic insertions will happen.

**Comment by oliver-sanders** [comment]
@hjoliver 

> can you explain in what sense exactly this is "part-baked": is there some circumstance in which you think it doesn't do the right thing, or is it just that you want to tidy up the implementation?

The implementation needs improving / tidying but the functionality should be there.

> I think this depends on how often (if any) "unexpected" automatic insertions will happen.

I haven't noticed any "unexpected" insertions yet. If the user were to move tasks across cycle points (with a reload) then it might **fail** to insert the task (there might be a workaround if there is a call for it).

File: lib/cylc/task_pool.py [hunk 1]
```diff
@@ -813,6 +819,234 @@ def pool_is_stalled(self):
...
+                        if sequence.is_on_sequence(point):
+                            flag = True
+                            break
+                    # ...if not insert at the first point on the tasks sequence
+                    # after the oldest active cycle point.
+                    if not flag:
+                        point = min([sequence.get_next_point(point) for sequence
+                                     in config.taskdefs[task.name].sequences])
+                        if not point:
+                            print ('ERROR: Could not determine cycle point to '
```
Comment (arjclark) [inline lib/cylc/task_pool.py]: presumably this should be a log.error or something to that effect?

**Comment by arjclark** [comment]
Not yet played around with this, but I think it should be an option to have cylc attempt an auto insert rather than default behaviour (e.g. for the isolate graphs they may want to be brought in manually at a specific time)

File: lib/cylc/task_pool.py (hunk 1 above)
Comment (oliver-sanders) [inline lib/cylc/task_pool.py]: Yep, this is proof of concept rather than merge-ready code.

**Comment by oliver-sanders** [comment]
Back with me whilst I think about implementation.

**Pull request closed without merging**